cpp/src/arrow/array.cc

```diff
@@ -1143,6 +1144,8 @@ struct ViewDataImpl {
 
     std::shared_ptr<ArrayData> out_data =
         ArrayData::Make(out_type, out_length, std::move(out_buffers), out_null_count);
+    out_data->dictionary = dictionary;
```

Comment:
Slightly simpler:
```suggestion
    if (out_type->id() == Type::DICTIONARY) {
      RETURN_NOT_OK(GetDictionaryView(*out_type, &out->dictionary));
    }
```

Reply:
Let's me see if this works, I was having to fiddle with the view-creation state machine

Reply:
This doesn't work.